lib/kafo/kafo_configure.rb

```diff
@@ -372,7 +395,7 @@ def is_value?(str)
     end
 
     def parse_app_arguments
-      self.class.app_options.each do |option|
+      self.class.declared_options.each do |option|
```

Comment:
I think this may introduce a regression where it stores all the params in `config.app` as well. Perhaps you can filter with `is_a?(AppOption)` on `declared_options`? Or introduce a compat helper since it's repeated above in `clamp_app_arguments`:
```ruby
def app_options
  declared_options.select { |option| option.is_a?(Kafo::AppOption::Definition) }
end
```

Reply:
First, *all* options are AppOption objects with this change. We can introduce a ModuleOption to be pedantic if we want but it's functionality would not vary. Secondly, there is no regression here due to the order in which things are processed and happen. The app arguments are parsed and handled before the params (module parameters are) dealt with so that they never end up in the config.app object.

https://github.com/theforeman/kafo/pull/299/files#diff-ef4f6d9e6e9975162ff8026e770e8787730a3547080b75534b8bdb411ff6efc2R145

Then at the end of initialization the module options are created:

https://github.com/theforeman/kafo/pull/299/files#diff-ef4f6d9e6e9975162ff8026e770e8787730a3547080b75534b8bdb411ff6efc2R151

```
[42, 51] in /root/kafo/lib/kafo/hook_context.rb
   42:     # determined during boot; this helper can be used in later hooks to determine whether the option exists.
   43:     #
   44:     # @param [Symbol, String] option
   45:     def app_option?(option)
   46:       require 'byebug';byebug
=> 47:       self.kafo.config.app.key?(option.to_sym)
   48:     end
   49: 
   50:     # @param [Symbol, String] option
   51:     #
(byebug) self.kafo.config.app
{:name=>"Foreman", :description=>"Default installation of Foreman", :enabled=>true, :log_dir=>"./_build/", :store_dir=>"", :log_name=>"foreman.log", :log_level=>:debug, :no_prefix=>false, :mapping=>{}, :answer_file=>"./config/foreman-answers.yaml", :installer_dir=>".", :module_dirs=>"./_build/modules", :colors=>true, :color_of_background=>:dark, :hook_dirs=>[], :custom=>{}, :facts=>{}, :low_priority_modules=>["foreman_proxy_plugin", "foreman_compute", "foreman_plugin"], :verbose_log_level=>"notice", :skip_puppet_version_check=>false, :hiera_config=>"./config/foreman-hiera.yaml", :parser_cache_path=>"./_build/parser_cache/foreman.yaml", :order=>["foreman", "foreman_proxy", "puppet"], :reset_data=>false, :detailed_exitcodes=>false, :clear_pulp_content=>false, :dont_save_answers=>false, :ignore_undocumented=>false, :interactive=>nil, :noop=>false, :profile=>false, :skip_checks_i_know_better=>false, :verbose=>true, :scenario=>nil, :disable_scenario=>nil, :enable_scenario=>nil, :list_scenarios=>nil, :force=>nil, :compare_scenarios=>nil, :migrations_only=>nil, :parser_cache=>nil, :help=>nil}
```

Reply:
How did you call it? Did you pass in `--mymodule-param value` while doing so? I'll try to set up a test env to actually try this patch (rather than just reading the patch).

Reply:
Correct. I did. If you look at the ordering in `initialize` method, the module options are not defined until *after* the app options are both defined and set.

App options defined: https://github.com/theforeman/kafo/pull/299/files#diff-ef4f6d9e6e9975162ff8026e770e8787730a3547080b75534b8bdb411ff6efc2R141
App options set: https://github.com/theforeman/kafo/pull/299/files#diff-ef4f6d9e6e9975162ff8026e770e8787730a3547080b75534b8bdb411ff6efc2R145
Module options defined and values set: 
  * https://github.com/theforeman/kafo/pull/299/files#diff-ef4f6d9e6e9975162ff8026e770e8787730a3547080b75534b8bdb411ff6efc2R151
  * https://github.com/theforeman/kafo/pull/299/files#diff-ef4f6d9e6e9975162ff8026e770e8787730a3547080b75534b8bdb411ff6efc2R357-R363

Reply:
To make things easier to test, I wrote https://github.com/ekohl/kafo-example-installer. Using this I confirmed this.

Reply:
> We can introduce a ModuleOption to be pedantic if we want but it's functionality would not vary.

I wonder if it makes sense in a follow up patch.